File: src/ros2_bridge/mappers/foxglove_msgs/point3_in_frame.rs
//! Typed mapper for `foxglove_msgs/msg/Point3InFrame`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn point3_in_frame_to_bus(msg: ros_env::foxglove_msgs::msg::Point3InFrame) -> crate::foxglove_msgs::msg::v1::Point3InFrame {
    crate::foxglove_msgs::msg::v1::Point3InFrame {
        timestamp: Some(crate::ros2_bridge::mappers::convert::time_to_timestamp(msg.timestamp)),
        frame_id: crate::ros2_bridge::mappers::convert::from_ros_string(msg.frame_id),
        point: Some(crate::ros2_bridge::mappers::foxglove_msgs::point3::point3_to_bus(msg.point)),
    }
}

pub(crate) fn point3_in_frame_to_ros(bus: crate::foxglove_msgs::msg::v1::Point3InFrame) -> ros_env::foxglove_msgs::msg::Point3InFrame {
    ros_env::foxglove_msgs::msg::Point3InFrame {
        timestamp: crate::ros2_bridge::mappers::convert::timestamp_to_time(bus.timestamp.unwrap_or_default()),
        frame_id: crate::ros2_bridge::mappers::convert::to_ros_string(bus.frame_id),
        point: crate::ros2_bridge::mappers::foxglove_msgs::point3::point3_to_ros(bus.point.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsPoint3InFrameMapper;

impl TypedTopicMapper for FoxgloveMsgsPoint3InFrameMapper {
    type Ros = ros_env::foxglove_msgs::msg::Point3InFrame;
    type Bus = crate::foxglove_msgs::msg::v1::Point3InFrame;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(point3_in_frame_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(point3_in_frame_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/point3_in_frame.py
"""Generated mapper for `foxglove_msgs/msg/Point3InFrame`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.point3 import point3_to_bus, point3_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import Point3InFrame as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def point3_in_frame_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.timestamp = _convert.time_to_timestamp(msg.timestamp)
    bus.frame_id = str(msg.frame_id)
    bus.point.CopyFrom(point3_to_bus(msg.point))
    return bus


def point3_in_frame_to_ros(bus):
    from foxglove_msgs.msg import Point3InFrame as RosMsg

    out = RosMsg()
    out.timestamp = _convert.timestamp_to_time(bus.timestamp)
    out.frame_id = str(bus.frame_id)
    out.point = point3_to_ros(bus.point)
    return out


class FoxgloveMsgsPoint3InFrameMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import Point3InFrame as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return point3_in_frame_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return point3_in_frame_to_ros(bus)
